doc/manpages/ccnlibtest.1
'\" t
.\"     Title: ccndlibtest
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 01/10/2014
.\"    Manual: \ \&
.\"    Source: \ \& 0.8.2
.\"  Language: English
.\"
.TH "CCNDLIBTEST" "1" "01/10/2014" "\ \& 0\&.8\&.2" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ccnlibtest \- Run scripted tests of the ccnx C library\&.
.SH "SYNOPSIS"
.sp
\fBccnlibtest\fR [\-h] [\-v] ( \fIcommand\fR | \fIfile\&.ccnb\fR ) \fI\&...\fR
.SH "DESCRIPTION"
.sp
The \fBccnlibtest\fR program is for running scripted tests of the ccnx C library\&.
.sp
The basic behavior of ccnlibtest is to use ccn_connect() to connect to a ccnd, and then to process its arguments, in order\&.
.SH "OPTIONS"
.PP
\fB\-h\fR
.RS 4
Print usage information and exit\&.
.RE
.PP
\fB\-v\fR
.RS 4
Be more verbose\&.
.RE
.SH "ARGUMENTS"
.sp
The arguments are processed in order\&.
.sp
An argument ending in \&.ccnb is assumed to name a file that contains a single ccnb\-encoded Interest or ContentObject\&. These files are read, and processed as follows:
.sp
A ContentObject is held in a local buffer pool\&. When triggered by the arrival of a matching interest, the ContentObject is sent, and also removed from the pool\&. There are up to 10 buffer pools available, and each buffer pool may contain up to 10 content objects\&. Initially pool 0 is the active pool\&.
.sp
An Interest is sent via ccn_express_interest()\&. If and when a matching ContentObject arrives, it will be placed into the pool that was active when the interest was sent\&.
.PP
\fBpool\fR \fIdigit\fR
.RS 4
Switch to the indicated buffer pool\&.
.RE
.PP
\fBprefix\fR \fIccnx:/uri\fR
.RS 4
Causes a prefix registration to happen, using the given name as the prefix\&. The registration is associated with the current buffer pool\&.
.RE
.PP
\fBflags\fR \fIregflgs\fR
.RS 4
Use
\fIregflgs\fR
as the flags for subsequent prefix registration\&. This should be expressed as a number\&. Refer to doc/technical/Registration for flag bits\&.
.RE
.PP
\fBrun\fR \fImillisec\fR
.RS 4
Run for the indicated number of milliseconds, or until the arrival of data\&.
.RE
.PP
\fBreconnect\fR \fIsecs\fR
.RS 4
Disconnect the ccn handle, sleep for
\fIsecs\fR
seconds and then reconnect\&. Used for testing disconnect/reconnect behavior\&.
.RE
.PP
\fBmincob\fR \fInum\fR
.RS 4
Fail if the current pool has fewer than
\fInum\fR
content objects\&.
.RE
.PP
\fBmaxcob\fR \fInum\fR
.RS 4
Fail if the current pool has more than
\fInum\fR
content objects\&.
.RE
.SH "SEE ALSO"
.sp
\fBccndsmoketest\fR(1)
.SH "AUTHOR"
.sp
Michael Plass


NEWS
Changes between 0.8.1 and 0.8.2:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/21 for more details on CCNx 0.8.2.

- This release features a refactored strategy layer, making it straightforward to add new
  strategies in a modular way.  Alongside the traditional "default" strategy there are
  two new examples, a "parallel" strategy that sends interests on all available faces
  and a "loadsharing" strategy that sends each arriving interest to just one of the
  available faces.  Strategies may be selected by prefix.  This selection may be
  done using new subcommands to the ccndc utility.

	* Feature #100929: Implement Junit test suites to facilitate bulk testing.
	* Feature #100983: strategy layer refactoring
	* Feature #100984: strategy labels on prefixes
	* Feature #100990: Restore SYNC_NOISY to INFO
	* Feature #100992: Add test-security to standard java tests
	* Feature #100993: Add test-versioning back into standard java tests
	* Feature #100994: Remove unused cruft from build.xml
	* Feature #101036: java sync should use Deque operations instead of Stack as recommended in the javadocs
	* Feature #101039: UDP mode for java netmanagers is not tested in standard testing
	* Feature #101053: ccncatchunks2 could use improved hole-filling and interest lifetime management
	* Bug #101030: build.xml setup of ccnr environment is incorrect
	* Bug #101033: Handling of extremely large values of MaxSuffixComponents is incorrect
	* Bug #101037: ccn_seqwriter library code should not limit the block size to 4096 bytes
	* Bug #101042: CcnFileProxy uses deprecated methods
	* Bug #101046: VLC 2.1.0 release necessitates change in plugin Makefile
	* Bug #101048: Do not rely on accept(2) returned addr for recognizing a unix-domain socket
	* Bug #101052: Unbreak objtree build target
	* Bug #101057: ccn_sign_content / ccn_chk_signing_params do not properly incorporate a timestamp in a template via ccn_signing_params
	* Bug #101058: ccndc in daemon mode does not exit cleanly when ccnd goes away
	* Bug #101059: testTimeouts a little too strict
	* Bug #101060: ccnd cache lookups should take a controllable amount of work
	* Bug #101061: ccnd minimum interest lifetime problem
	* Bug #101062: some ccnd logging missing check for logging enabled before calling expensive log procedure
	* Support #101047: Prepare for ccnx-0.8.2 release

Changes between 0.8.0 and 0.8.1:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/20 for more details on CCNx 0.8.1.

- This release fixes a number of bugs, including a serious memory leak in ccnd.

	* Bug #101029: unbounded test log file growth
	* Bug #101032: ccnr should start at specified segment on checked start write
	* Bug #101034: No obvious documentation for how to start ccnr exists on website
	* Bug #101035: small leak in ccn_fetch should a realloc() fail
	* Bug #101038: Inconsistent encoding of zero-length name components
	* Bug #101043: ccnr r_store_lookup_backwards can fail
	* Bug #101044: valgrind reports memory leak in 0.8.0 ccnd
	* Bug #101049: ccnd can remove newly arrived content too quickly
	* Support #101045 Prepare for ccnx-0.8.1 release

Changes between 0.7.2 and 0.8.0:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/17 for more details on CCNx 0.8.0.

- The use of symmetric key HMACs in place of public/private key signatures
  is now an option in both libraries.  In situations where it can be used,
  it provides a much higher performance alternative.
- The content store in ccnd has been improved.  It now uses a flatname
  representation for indexing, in the same way that ccnr does.  Also, the
  underlying skiplist implementation has moved into the library, so that
  applications may take advantage of it in the future.  In the process,
  some old ccnd bugs relating to the handling of extreme values of
  FreshnessSeconds were fixed.
- The more readable escaping convention in URIs is now the default.
  If for whatever reason you prefer to use only percent-escaping in URIs, use

	CFLAGS='-g -DCCN_URI_DEFAULT_ESCAPE=4' ./configure

  before typing make.  Note that both forms of escaping are accepted as input.
- An example rc startup script is now included, but needs to be installed by
  hand.  It is coded for Ubuntu Linux, but should be easily adaptable to other
  systems.
- This release drops support for Java JDK 1.5.

  * Issues Resolved:
	* Bug #100118: ccnd does not handle FreshnessSeconds=0 correctly.
	* Bug #100119: ccnd should handle large values of FreshnessSeconds
	* Bug #100558: test-security does not pass
	* Bug #100897: timeout values  should be consistently long throughout java code 
	* Bug #100966: Incorrect processing of CCNS_SYNC_SCOPE=3
	* Bug #100977: Problem with verify in CCNInputStream when trying to retrieve key
	* Bug #100987: Non-portable use of local keyword in Bourne shell scripts
	* Bug #100998: VLC 2.1.x block_New definition removed
	* Bug #100999: java BLOCK_AFTER_FIRST_SEGMENT doesn't work in all cases
	* Bug #101000: sync_diff code should reset differencing data when aborting or completing a comparison
	* Bug #101004: QueuedContentHandler has a sync error
	* Bug #101005: CCNx Android Services build out of sync with API changes to CCNx C-lib
	* Bug #101007: ccn_resolve_version() should distinguish between no response and finding an unversioned object.
	* Bug #101010: ccnseqwriter command busy-loops when its ccnd goes away
	* Bug #101011: Compile warning when building new symmetric key code with openssl 0.9.8
	* Bug #101012: Improve recovery from node fetch timeout in java sync
	* Bug #101013: node fetch timeout in java sync implementation should be longer and settable

csrc/cmd/ccnlibtest.c
/**
 * @file cmd/ccnlibtest.c
 */
/*
 * A CCNx program.
 *
 * Copyright (C) 2009, 2010, 2012 Palo Alto Research Center, Inc.
 *
 * This work is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License version 2 as published by the
 * Free Software Foundation.
 * This work is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details. You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */
 
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <unistd.h>

#include <ccn/ccn.h>
#include <ccn/reg_mgmt.h>
#include <ccn/uri.h>

int verbose;

void
printraw(const void *r, int n)
{
    int i, l;
    const unsigned char *p = r;
    
    if (verbose == 0)
        return;
    while (n > 0) {
        l = (n > 40 ? 40 : n);
        for (i = 0; i < l; i++)
            printf(" %c", (' ' <= p[i] && p[i] <= '~') ? p[i] : '.');
        printf("\n");
        for (i = 0; i < l; i++)
            printf("%02X", p[i]);
        printf("\n");
        p += l;
        n -= l;
    }
}
/* Use some static data for this simple program */
static unsigned char rawbuf[65536];
static ssize_t rawlen;

#define N_POOLS 10
#define MINI_STORE_LIMIT 10
struct mini_store {
    struct ccn_closure me;
    struct ccn_charbuf *cob[MINI_STORE_LIMIT];
};
static struct ccn_closure incoming_content_action[N_POOLS];
static struct mini_store store[N_POOLS];

int
add_to_pool(int pool, const unsigned char *r, size_t n)
{
    int i, j;
    struct ccn_charbuf **coba;
    
    coba = store[pool].cob;
    for (i = 0, j = 0; i < MINI_STORE_LIMIT; i++) {
        if (coba[i] != NULL)
            coba[j++] = coba[i];
    }
    for (i = j; i < MINI_STORE_LIMIT; i++)
        coba[i] = NULL;
    if (j < MINI_STORE_LIMIT) {
        coba[j] = ccn_charbuf_create();
        ccn_charbuf_append(coba[j], r, n);
        return(j + 1);
    }
    return(-1);
}

int
n_pool(int pool)
{
    int i, n;
    struct ccn_charbuf **coba;
    
    coba = store[pool].cob;
    for (i = 0, n = 0; i < MINI_STORE_LIMIT; i++)
        if (coba[i] != NULL)
            n++;
    return(n);
}

enum ccn_upcall_res
incoming_content(struct ccn_closure *selfp,
                 enum ccn_upcall_kind kind,
                 struct ccn_upcall_info *info)
{
    if (kind == CCN_UPCALL_FINAL)
        return(CCN_UPCALL_RESULT_OK);
    if (kind == CCN_UPCALL_INTEREST_TIMED_OUT)
        return(CCN_UPCALL_RESULT_REEXPRESS);
    if (kind != CCN_UPCALL_CONTENT && kind != CCN_UPCALL_CONTENT_UNVERIFIED)
        return(CCN_UPCALL_RESULT_ERR);
    printf("Got content matching %d components:\n", info->pi->prefix_comps);
    printraw(info->content_ccnb, info->pco->offset[CCN_PCO_E]);
    add_to_pool(selfp->intdata, info->content_ccnb, info->pco->offset[CCN_PCO_E]);
    return(CCN_UPCALL_RESULT_OK);
}

int
cob_matches(struct ccn_upcall_info *info, struct ccn_charbuf *cob)
{
    int ans;
    
    ans = ccn_content_matches_interest(cob->buf, cob->length, 1, NULL,
                                       info->interest_ccnb,
                                       info->pi->offset[CCN_PI_E],
                                       info->pi);
    return(ans);
}

enum ccn_upcall_res
outgoing_content(struct ccn_closure *selfp,
                 enum ccn_upcall_kind kind,
                 struct ccn_upcall_info *info)
{
    struct mini_store *md;
    struct ccn_charbuf *cob = NULL;
    int i;
    int res = 0;
    int which;
    
    md = selfp->data;
    which = md->me.intdata;
    if (kind == CCN_UPCALL_FINAL) {
        fprintf(stderr, "CCN_UPCALL_FINAL for store %d\n", which);
        for (i = 0; i < MINI_STORE_LIMIT; i++)
            ccn_charbuf_destroy(&md->cob[i]);
        return(CCN_UPCALL_RESULT_OK);
    }
    printf("Store %d got interest matching %d components, kind = %d",
           which, info->matched_comps, kind);
    /* Look through our little pile of content and send one that matches */
    if (kind == CCN_UPCALL_INTEREST) {
        for (i = 0; i < MINI_STORE_LIMIT; i++) {
            cob = md->cob[i];
            if (cob != NULL && cob_matches(info, cob)) {
                res = ccn_put(info->h, cob->buf, cob->length);
                if (res == -1) {
                    fprintf(stderr, "... error sending data\n");
                    return(CCN_UPCALL_RESULT_ERR);
                }
                else {
                    printf("... sent my content:\n");
                    printraw(cob->buf, cob->length);
                    ccn_charbuf_destroy(&md->cob[i]);
                    return(CCN_UPCALL_RESULT_INTEREST_CONSUMED);
                }
            }
        }
        printf("... no match\n");
    }
    else
        printf("\n");
    return(CCN_UPCALL_RESULT_ERR);
}

#define USAGE "ccnlibtest [-hv] (pool n | flags x | prefix uri | reconnect secs | run millis | file.ccnb) ..."

void
usage(void)
{
    fprintf(stderr, USAGE "\n");
    exit(1);
}

int
main(int argc, char **argv)
{
    char *arg = NULL;
    struct ccn *ccnH = NULL;
    struct ccn_parsed_interest interest = {0};
    struct ccn_charbuf *c = ccn_charbuf_create();
    struct ccn_charbuf *templ = ccn_charbuf_create();
    struct ccn_indexbuf *comps = ccn_indexbuf_create();
    int i;
    int millis = 0;
    int secs = 0;
    int opt;
    int pool = 0;
    int regflgs = (CCN_FORW_CHILD_INHERIT | CCN_FORW_ACTIVE);
    int res;
    int status = 0;
    int val = 0;
    
    while ((opt = getopt(argc, argv, "hv")) != -1) {
        switch (opt) {
            default:
            case 'h':
                usage();
                break;
            case 'v':
                verbose++;
                break;
        }
    }
    argc -= optind;
    argv += optind;
    ccnH = ccn_create();
    if (ccn_connect(ccnH, NULL) == -1) {
        ccn_perror(ccnH, "ccn_connect");
        exit(1);
    }
    for (i = 0; i < N_POOLS; i++) {
        store[i].me.p = &outgoing_content;
        store[i].me.data = &store[i];
        store[i].me.intdata = i;
        incoming_content_action[i].p = &incoming_content;
        incoming_content_action[i].intdata = i;
    }
    for (i = 0; i < argc; i++) {
        arg = argv[i];
        if (0 == strcmp(arg, "reconnect")) {
            if (argv[i+1] == NULL)
                usage();
            secs = atoi(argv[i+1]);
            if (secs <= 0 && strcmp(argv[i+1], "0") != 0)
                usage();
            i++;
            ccn_disconnect(ccnH);
            sleep(secs);
            if (ccn_connect(ccnH, NULL) == -1) {
                ccn_perror(ccnH, "ccn_connect");
                exit(1);
            }
            continue;
        }
        if (0 == strcmp(arg, "pool")) {
            if (argv[i+1] == NULL)
                usage();
            pool = argv[i+1][0] - '0';
            if (argv[i+1][1] || pool < 0 || pool >= N_POOLS)
                usage();
            fprintf(stderr, "Pool %d\n", pool);
            i++;
            continue;
        }
        if (0 == strcmp(arg, "prefix")) {
            if (argv[i+1] == NULL)
                usage();
            c->length = 0;
            res = ccn_name_from_uri(c, argv[i+1]);
            if (res < 0)
                usage();
            fprintf(stderr, "Prefix ff=%#x %s pool %d\n",
                    regflgs, argv[i+1], pool);
            if (store[pool].me.intdata != pool) {
                abort();
            }
            res = ccn_set_interest_filter_with_flags(ccnH, c, &store[pool].me, regflgs);
            if (res < 0) {
                ccn_perror(ccnH, "ccn_set_interest_filter_with_flags");
                status = 1;
            }
            res = ccn_run(ccnH, 2);
            if (res < 0)
                break;
            i++;
            continue;
        }
        if (0 == strcmp(arg, "flags")) {
            if (argv[i+1] == NULL)
                usage();
            regflgs = atoi(argv[i+1]);
            if (regflgs <= 0 && strcmp(argv[i+1], "0") != 0)
                usage();
            i++;
            continue;
        }
        if (0 == strcmp(arg, "mincob")) {
            if (argv[i+1] == NULL)
                usage();
            val = atoi(argv[i+1]);
            if (val <= 0 && strcmp(argv[i+1], "0") != 0)
                usage();
            i++;
            if (n_pool(pool) < val) {
                fprintf(stderr, "Pool %d has %d cobs, expected at least %d\n",
                        pool, n_pool(pool), val);
                exit(1);
            }
            continue;
        }
        if (0 == strcmp(arg, "maxcob")) {
            if (argv[i+1] == NULL)
                usage();
            val = atoi(argv[i+1]);
            if (val <= 0 && strcmp(argv[i+1], "0") != 0)
                usage();
            i++;
            if (n_pool(pool) > val) {
                fprintf(stderr, "Pool %d has %d cobs, expected at most %d\n",
                        pool, n_pool(pool), val);
                exit(1);
            }
            continue;
        }
        if (0 == strcmp(arg, "run")) {
            if (argv[i+1] == NULL)
                usage();
            millis = atoi(argv[i+1]);
            if (millis <= 0 && strcmp(argv[i+1], "0") != 0)
                usage();
            i++;
            res = ccn_run(ccnH, millis);
            if (res < 0) {
                ccn_perror(ccnH, "ccn_run");
                exit(1);
            }
            continue;
        }
        close(0);
        res = open(arg, O_RDONLY);
        if (res != 0) {
            perror(arg);
            exit(1);
        }
        fprintf(stderr, "Reading %s ... ", arg);
        rawlen = read(0, rawbuf, sizeof(rawbuf));
        if (rawlen < 0) {
            perror("skipping");
            // XXX - status
            continue;
        }
        // XXX - Should do a skeleton check before parse
        res = ccn_parse_interest(rawbuf, rawlen, &interest, NULL);
        if (res >= 0) {
            size_t name_start = interest.offset[CCN_PI_B_Name];
            size_t name_size = interest.offset[CCN_PI_E_Name] - name_start;
            templ->length = 0;
            ccn_charbuf_append(templ, rawbuf, rawlen);
            fprintf(stderr, "Expressing interest with %d name components\n", res);
            c->length = 0;
            ccn_charbuf_append(c, rawbuf + name_start, name_size);
            // XXX - res is currently ignored
            ccn_express_interest(ccnH, c, &(incoming_content_action[pool]), templ);
        }
        else {
            struct ccn_parsed_ContentObject obj = {0};
            int try;
            res = ccn_parse_ContentObject(rawbuf, rawlen, &obj, comps);
            if (res >= 0) {
                for (try = 0; try < 5; try++) {
                    res = add_to_pool(pool, rawbuf, rawlen);
                    if (res >= 0) {
                        fprintf(stderr, "Added to pool %d\n", pool);
                        break;
                    }
                    if (try == 5) {
                        fprintf(stderr, "No buffer for %s\n", arg);
                        status = 1;
                        break;
                    }
                    fprintf(stderr, "Pool %d full - wait for drain\n", pool);
                    if (ccn_run(ccnH, 1000) < 0)
                        break;
                }
                res = ccn_run(ccnH, 10);
            }
            else {
                fprintf(stderr, "What is that?\n");
                status = 1;
            }
        }
        res = ccn_run(ccnH, 10);
        if (res < 0) {
            ccn_perror(ccnH, "oops");
            status = 1;
        }
    }
    res = ccn_run(ccnH, 10);
    if (res < 0)
        status = 1;
    ccn_destroy(&ccnH);
    exit(status);
}
